File: src/jastudio/note/note_constants.py
# noinspection PyUnusedName
from __future__ import annotations

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from typed_linq_collections.collections.q_set import QSet


class Builtin(Slots):
    # noinspection PyUnusedName
    Tag: str = "tag"
    Note: str = "note"
    Deck: str = "deck"
    Card: str = "card"

class MyNoteFields(Slots):
    question: str = "Q"
    answer: str = "A"

class ImmersionKitSentenceNoteFields(Slots):
    audio: str = "Audio Sentence"
    id: str = "ID"
    screenshot: str = "Screenshot"
    reading: str = "Reading"
    answer: str = "English"
    question: str = "Expression"

class SentenceNoteFields(Slots):
    reading: str = "Reading"
    id: str = "ID"
    active_question: str = MyNoteFields.question
    source_question: str = "source_question"
    source_comments: str = "Comments"
    user_comments: str = "__comments"
    user_question: str = "__question"
    active_answer: str = MyNoteFields.answer
    source_answer: str = "source_answer"
    user_answer: str = "__answer"
    parsing_result: str = "__parsing_result"
    audio: str = "Audio Sentence"
    screenshot: str = "Screenshot"
    configuration: str = "__configuration"

class CardTypes(Slots):
    reading: str = "Reading"
    listening: str = "Listening"

class NoteTypes(Slots):
    immersion_kit: str = "Immersion Kit Sentence"
    Kanji: str = "_Kanji"
    Vocab: str = "_Vocab"
    Sentence: str = "_japanese_sentence"

    ALL: QSet[str] = QSet((Kanji, Vocab, Sentence))

class NoteFields(Slots):
    note_id: str = "nid"

    class VocabNoteType(Slots):
        class Card(Slots):
            Reading: str = CardTypes.reading
            Listening: str = CardTypes.listening

    class SentencesNoteType(Slots):
        class Card(Slots):
            Reading: str = CardTypes.reading
            Listening: str = CardTypes.listening

    class Kanji(Slots):
        question: str = MyNoteFields.question
        active_answer: str = MyNoteFields.answer
        source_answer: str = "source_answer"
        user_answer: str = "__answer"
        Reading_On: str = "Reading_On"
        Reading_Kun: str = "Reading_Kun"
        Reading_Nan: str = "__reading_Nan"
        Radicals: str = "Radicals"
        Source_Meaning_Mnemonic: str = "Meaning_Mnemonic"
        Meaning_Info: str = "Meaning_Info"
        Reading_Mnemonic: str = "Reading_Mnemonic"
        Reading_Info: str = "Reading_Info"
        PrimaryVocab: str = "__primary_Vocab"
        Audio__: str = "__audio"

        user_mnemonic: str = "__mnemonic"
        user_similar_meaning: str = "__similar_meaning"
        related_confused_with: str = "__confused_with"

    class Vocab(Slots):
        matching_rules: str = "__matching_rules"
        related_vocab: str = "__related_vocab"
        sentence_count: str = "sentence_count"
        question: str = MyNoteFields.question
        active_answer: str = MyNoteFields.answer
        source_answer: str = "source_answer"
        user_answer: str = "__answer"
        user_explanation: str = "__explanation"
        user_explanation_long: str = "__explanation_long"
        user_compounds: str = "__compounds"
        user_mnemonic: str = "__mnemonic"
        Reading: str = "Reading"
        parts_of_speech: str = "TOS"
        source_mnemonic: str = "source_mnemonic"
        Audio_b: str = "Audio_b"
        Audio_g: str = "Audio_g"
        Audio_TTS: str = "Audio_TTS"

        Forms: str = "F"
        source_reading_mnemonic: str = "source_reading_mnemonic"
        Homophones: str = "Homophones"
        ParsedTypeOfSpeech: str = "ParsedTypeOfSpeech"

class Mine(Slots):
    app_name: str = "JA-Studio"
    app_still_loading_message: str = f"{app_name} still loading, the view will refresh when done..."
    VocabPrefixSuffixMarker: str = "〜"

File: src/src_dotnet/JAStudio.Core/Note/NoteConstants.cs
using System;
using System.Collections.Generic;
using JAStudio.Core.Note.Sentences;
using JAStudio.Core.Note.Vocabulary;

namespace JAStudio.Core.Note;

public static class MyNoteFields
{
   public const string Question = "Q";
   public const string Answer = "A";
   public const string JasNoteId = "jas_note_id";
}

public static class SentenceNoteFields
{
   public const string Reading = "Reading";
   public const string Id = "ID";
   public const string ActiveQuestion = MyNoteFields.Question;
   public const string SourceQuestion = "source_question";
   public const string SourceComments = "Comments";
   public const string UserComments = "__comments";
   public const string UserQuestion = "__question";
   public const string ActiveAnswer = MyNoteFields.Answer;
   public const string SourceAnswer = "source_answer";
   public const string UserAnswer = "__answer";
   public const string ParsingResult = "__parsing_result";
   public const string JanomeTokens = "__janome_tokens";
   public const string Audio = "Audio Sentence";
   public const string Screenshot = "Screenshot";
   public const string Configuration = "__configuration";
}

public static class CardTypes
{
   public const string Reading = "Reading";
   public const string Listening = "Listening";
}

public static class NoteTypes
{
   public const string ImmersionKit = "Immersion Kit Sentence";
   public const string Kanji = "_Kanji";
   public const string Vocab = "_Vocab";
   public const string Sentence = "_japanese_sentence";

   public static readonly HashSet<string> All = [Kanji, Vocab, Sentence];

   public static readonly IReadOnlyList<string> AllList = [Kanji, Vocab, Sentence];

   public static string FromType(Type noteType) => Map[noteType];

   public static Func<Guid, NoteId> IdFactoryFromType(Type noteType) => IDFactories[noteType];

   public static Func<Guid, NoteId> IdFactoryFromName(string noteTypeName) => IDFactoriesByName[noteTypeName];

   //create a dictionary from type to string for our supported type her
   static readonly Dictionary<Type, string> Map = new()
                                                  {
                                                     [typeof(KanjiNote)] = Kanji,
                                                     [typeof(VocabNote)] = Vocab,
                                                     [typeof(SentenceNote)] = Sentence,
                                                  };

   static readonly Dictionary<Type, Func<Guid, NoteId>> IDFactories = new()
                                                                      {
                                                                         [typeof(KanjiNote)] = g => new KanjiId(g),
                                                                         [typeof(VocabNote)] = g => new VocabId(g),
                                                                         [typeof(SentenceNote)] = g => new SentenceId(g),
                                                                      };

   static readonly Dictionary<string, Func<Guid, NoteId>> IDFactoriesByName = new()
                                                                              {
                                                                                 [Kanji] = g => new KanjiId(g),
                                                                                 [Vocab] = g => new VocabId(g),
                                                                                 [Sentence] = g => new SentenceId(g),
                                                                              };
}

public static class NoteFieldsConstants
{
   public const string NoteId = "nid";

   public static class VocabNoteType
   {
      public static class Card
      {
         public const string Reading = CardTypes.Reading;
         public const string Listening = CardTypes.Listening;
      }
   }

   public static class SentencesNoteType
   {
      public static class Card
      {
         public const string Reading = CardTypes.Reading;
         public const string Listening = CardTypes.Listening;
      }
   }

   public static class Kanji
   {
      public const string Question = MyNoteFields.Question;
      public const string ActiveAnswer = MyNoteFields.Answer;
      public const string SourceAnswer = "source_answer";
      public const string UserAnswer = "__answer";
      public const string ReadingOn = "Reading_On";
      public const string ReadingKun = "Reading_Kun";
      public const string ReadingNan = "__reading_Nan";
      public const string Radicals = "Radicals";
      public const string SourceMeaningMnemonic = "Meaning_Mnemonic";
      public const string MeaningInfo = "Meaning_Info";
      public const string ReadingMnemonic = "Reading_Mnemonic";
      public const string ReadingInfo = "Reading_Info";
      public const string PrimaryVocab = "__primary_Vocab";
      public const string Audio = "__audio";
      public const string PrimaryReadingsTTSAudio = "_primary_readings_tts_audio";
      public const string References = "__references";
      public const string UserMnemonic = "__mnemonic";
      public const string UserSimilarMeaning = "__similar_meaning";
      public const string RelatedConfusedWith = "__confused_with";
      public const string Image = "_image";
   }

   public static class Vocab
   {
      public const string MatchingRules = "__matching_rules";
      public const string RelatedVocab = "__related_vocab";
      public const string SentenceCount = "sentence_count";
      public const string Question = MyNoteFields.Question;
      public const string ActiveAnswer = MyNoteFields.Answer;
      public const string SourceAnswer = "source_answer";
      public const string UserAnswer = "__answer";
      public const string UserExplanation = "__explanation";
      public const string UserExplanationLong = "__explanation_long";
      public const string UserCompounds = "__compounds";
      public const string UserMnemonic = "__mnemonic";
      public const string Reading = "Reading";
      public const string PartsOfSpeech = "TOS";
      public const string SourceMnemonic = "source_mnemonic";
      public const string AudioB = "Audio_b";
      public const string AudioG = "Audio_g";
      public const string AudioTTS = "Audio_TTS";
      public const string Kanji = "__kanji";
      public const string Forms = "F";
      public const string SourceReadingMnemonic = "source_reading_mnemonic";
      public const string TechnicalNotes = "__technical_notes";
      public const string References = "__references";
      public const string Image = "Image";
      public const string UserImage = "__image";
   }
}

public static class Builtin
{
   public const string Tag = "tag";
   public const string Note = "note";
   public const string Deck = "deck";
   public const string Card = "card";
}

public static class Mine
{
   public const string AppName = "JA-Studio";

   // ReSharper disable once UnusedMember.Global - Used from Python (dotnet_rendering_content_renderer_anki_shim.py, common.py)
   public static readonly string AppStillLoadingMessage = $"{AppName} still loading, the view will refresh when done...";
   public const string VocabPrefixSuffixMarker = "〜";
}
